lib/complete-provider.coffee

```diff
@@ -53,11 +58,19 @@ class CompleteProvider extends Provider
     'then', 'where'
   ]
 
-  initialize: (@editorView, @manager) ->
+  constructor: (@editor, @manager) ->
+    if (@editor.haskellCompletionProvider && @editor.haskellCompletionProvider != this)
+      # shouldn't happen, but log just in case
+      console.log "weird: editor already has a haskell completion provider"
+      @editor.haskellCompletionProvider.dispose()
```

Comment:
Gotcha. Well then yeah, have a single provider and a cache, and compare options.editor with the cache. Weak map seems nice, cos it would drop the cache of the editor is destroyed.